component/loki/source/kubernetes/kubetail/tailer.go

```diff
@@ -0,0 +1,323 @@
+	defer level.Info(t.log).Log("msg", "tailer exited")
+
+	bo := backoff.New(ctx, retailBackoff)
+
+	handler := loki.NewEntryMutatorHandler(t.opts.Handler, func(e loki.Entry) loki.Entry {
+		// A log line got read, we can reset the backoff period now.
+		bo.Reset()
+		return e
+	})
+	defer handler.Stop()
+
+	for bo.Ongoing() {
+		err := t.tail(ctx, handler)
+		if err == nil {
+			terminated, err := t.containerTerminated(ctx)
+			if terminated {
+				// The container shut down and won't come back; we can stop tailing it.
+				return
+			} else if err != nil {
+				level.Warn(t.log).Log("msg", "could not determine if container terminated; will retry tailing", "err", err)
+			}
+		}
+
+		if err != nil {
+			t.target.Report(time.Now().UTC(), err)
+			level.Warn(t.log).Log("msg", "tailer stopped; will retry", "err", err)
+		}
+		bo.Wait()
+	}
+}
+
+func (t *tailer) tail(ctx context.Context, handler loki.EntryHandler) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var (
+		key           = t.target.NamespacedName()
+		containerName = t.target.ContainerName()
+
+		positionsEnt = entryForTarget(t.target)
+	)
+
+	var lastReadTime time.Time
+
+	if offset, err := t.opts.Positions.Get(positionsEnt.Path, positionsEnt.Labels); err != nil {
+		level.Warn(t.log).Log("msg", "failed to load last read offset", "err", err)
+	} else {
+		lastReadTime = time.UnixMicro(offset)
+	}
+
+	// If the last entry for our target is after the positions cache, use that
+	// instead.
+	if lastEntry := t.target.LastEntry(); lastEntry.After(lastReadTime) {
+		lastReadTime = lastEntry
+	}
+
+	var offsetTime *metav1.Time
+	if !lastReadTime.IsZero() {
+		offsetTime = &metav1.Time{Time: lastReadTime}
+	}
+
+	req := t.opts.Client.CoreV1().Pods(key.Namespace).GetLogs(key.Name, &corev1.PodLogOptions{
+		Follow:     true,
+		Container:  containerName,
+		SinceTime:  offsetTime,
+		Timestamps: true, // Should be forced to true so we can parse the original timestamp back out.
+	})
+	stream, err := req.Stream(ctx)
+	if err != nil {
+		return err
+	}
+	go func() {
+		<-ctx.Done()
+		_ = stream.Close()
+	}()
+
+	level.Info(t.log).Log("msg", "opened log stream", "start time", lastReadTime)
+
+	ch := handler.Chan()
+	reader := bufio.NewReader(stream)
+
+	for {
+		line, err := reader.ReadString('\n')
+
+		// Try processing the line before handling the error, since data may still
+		// be returned alongside an EOF.
+		if len(line) != 0 {
+			entryTimestamp, entryLine := parseKubernetesLog(line)
+			if !entryTimestamp.After(lastReadTime) {
+				continue
+			}
+			lastReadTime = entryTimestamp
+
+			entry := loki.Entry{
+				Labels: t.lset,
+				Entry: logproto.Entry{
+					Timestamp: entryTimestamp,
+					Line:      entryLine,
+				},
+			}
+
+			select {
+			case <-ctx.Done():
+				return nil
+			case ch <- entry:
+				// Save position after it's been sent over the channel.
+				t.opts.Positions.Put(positionsEnt.Path, positionsEnt.Labels, entryTimestamp.UnixMicro())
```

Comment:
UnixMicro() by default returns the local time, should we store it in UTC?

I couldn't find a definitive answer as to whether Kubernetes clusters can set their timezones. This might be a footgun for someone who chooses to run their pods on different timezones, but I think we should standardize around the common usecase.

FWIW [RFC3339](https://www.rfc-editor.org/rfc/rfc3339#section-4.1) does state that "This specification does not cater to local time zone rules" encouraging UTC, but it _does_ allow offsets.

Reply:
Are you sure it returns the local time?

```
// UnixMicro returns t as a Unix time, the number of microseconds elapsed since
// January 1, 1970 UTC. The result is undefined if the Unix time in
// microseconds cannot be represented by an int64 (a date before year -290307 or
// after year 294246). The result does not depend on the location associated
// with t.
```

I can change this to `entryTimestamp.UTC().UnixMicro()` just to be safe though.

Reply:
Actually, I think we should store it in whatever format we get back from Kubernetes. Just it case it *does* change, we don't want to store a different time in the positions file and then load that different time back in for the next tail (since at that point, we wouldn't know how to convert it back to the local time Kubernetes recognizes)

Reply:
Whoops, I was looking at the docs for the UnixMicro function, not the method https://pkg.go.dev/time#UnixMicro 🤷